airflow/jobs/triggerer_job.py

```diff
@@ -17,26 +17,175 @@
+            if isinstance(h, FileTaskHandler) and not supports_triggerer(h):
+                warnings.warn(
+                    f"Handler {h.__class__.__name__} does not support "
+                    "individual trigger logging. Please check the release notes "
+                    "for your provider to see if a newer version supports "
+                    "individual trigger logging."
+                )
+            if supports_triggerer(h):
+                return h
+
+    def find_suitable_task_handler():
+        # check root logger then check airflow.task to see if a handler
+        # suitable for use with TriggerHandlerWrapper (has supports_triggerer
+        # attr, likely inherits from FileTaskHandler)
+        h = get_task_handler_from_logger(root_logger)
+        if not h:
+            # try to use handler configured from airflow task
+            logger.debug("No task logger configured for root logger; trying `airflow.task`.")
+            h = get_task_handler_from_logger(logging.getLogger("airflow.task"))
+            if h:
+                logger.debug("Using logging configuration from `airflow.task`")
+        if not h:
+            warnings.warn("Could not find log handler suitable for individual trigger logging.")
+            return None
+        return h
+
+    def filter_trigger_logs_from_other_root_handlers(new_hdlr):
+        # we add context vars to log records emitted for individual triggerer logging
+        # we want these records to be processed by our special trigger handler wrapper
+        # but not by any other handlers, so we filter out these messages from
+        # other handlers by adding DropTriggerLogsFilter
+        # we could consider only adding this filter to the default console logger
+        # so as to leave other custom handlers alone
+        for h in root_logger.handlers:
+            if h is not new_hdlr:
+                h.addFilter(DropTriggerLogsFilter())
+
+    def add_handler_wrapper_to_root(base_handler):
+        # first make sure we remove from root logger if it happens to be there
+        # it could have come from root or airflow.task, but we only need
+        # to make sure we remove from root, since messages will not flow
+        # through airflow.task
+        if base_handler in root_logger.handlers:
+            root_logger.removeHandler(base_handler)
+
+        h = TriggererHandlerWrapper(base_handler=base_handler, level=base_handler.level)
+        # just extra cautious, checking if user manually configured it there
+        if h not in root_logger.handlers:
+            root_logger.addHandler(h)
+        return h
+
+    root_logger = logging.getLogger()
+    task_handler = find_suitable_task_handler()
+    if not task_handler:
+        return None
+    if TYPE_CHECKING:
+        assert isinstance(task_handler, FileTaskHandler)
+    wrapper_handler = add_handler_wrapper_to_root(task_handler)
+    filter_trigger_logs_from_other_root_handlers(wrapper_handler)
```

Comment:
@ashb hopefully i made this a little easier to digest by shoving the logic into functions

Reply:
It's quite clear. BTW. This is an interesting problem to solve with blocking I/O in log handlers.

However, I just wonder if this is really the kind of complexity we need here (and yes.

I am sorry for being late in the game here as I had other stuff to take a look at ) so maybe I do not understand something and maybe I miss a crucial piece of information, apologies If I do, but I think we can do it slightly simpler, and in the way that we will get full support for any (even completely custom) streaming log handlers from day one.

What would be nice there is that the "Trigger logs" tab would only be needed if you do not use remote logging or when your remote logging does not support streaming during task execution (like s3). In case of streaming loggers instead, we will get the logs interleaved with the regular task logs - which is much better, because then you will see them in one "stream" (barring occasional races which are inevitable in a distributed system - but with proper time synchronisation between nodes and support from log streaming might be limited to whatever precision the logging solution supports (ms/us etc). 

I think for all cases where streaming logs **just works** during code execution, we would not have to do anything - they would be handed out-of-the-box. It would also utilize most of the other code in this PR - just setting up the handlers woudl be simpler, and it would support all streaming handers out of the box with 0 modifications to them.

The approach I think we could use in this case is as follows:

1) let all trigger code use "airflow.trigger" logger - which would have always a single handler, that would only do non-blocking calls - forwarding the logs to an in-memory queue, including context of the task instance that is the "cause" of the log.

2) Have a separate non-async-io compatible process that reads from the in-memory queue and .... log it to "airflow.task" logger - pretending they are a special "worker". It could also serve the logs in the very same way as the other workers (and handle the Trigger logs tab). No manipulation of handlers is needed in this case. This separate process would have just the same set of loggers any other airflow process has. The only difference is that it would have to pass the context to "set_context" method of the configured logger before writing a log entry. 

3) #ither now or in the future this approach could be quite optimized so that the context does not switch too frequently (this is the main problem I see we have when one triggerer handles multiple tasks). For example we should allow N parallel threads (or processes) to write logs - each with its own queue based on hash of the dag_id, task_id, run_id. Then each such thread/process could do its own micro-batching for writing the logs (with slight delay) and sort them according to dag_id, task_id, run_id before writing, so that multiple log entries for the same task run could be written together in a batch with just a single "set_context" call.

The benefit of doing it this way, all (including custom) streaming handlers will be handled out-of-the-box from day 0 - and neither the users nor us will have to make a single change to those handlers to make them work nicely - including the fact that the trigger log entries will be nicely interleaved with the "regular" task logs in case of streaming solutions, which is exactly what our users expect. 

I believe that should be about it. No manipulation with handlers, no code for custom, streaming handlers, we can implement optimisation for context switching, same code handling local logs and serving them as for the streaming solutions.

Reply:
Gave the comment a quick skim will read more closely later but...

I originally interleaved but I thought that separate tab was actually better. Because it's if course easier to see where your messages are coming from this way, and easier to find what you are looking for e.g. trigger logs vs task.  But I could be convinced otherwise.

Re the solution for blocking io. This is a good practice in general and is more user friendly since it makes blocking io in handlers a non problem.  If you Google it is recommended in official python docs.

And see blog post from Martijn Peters I think his name is ...?

Your suggestions seem to describe pretty accurately what I am doing already. Am continuing to try to refine a bit to reduce the adjustments required in existing handlers. E.g. locally I have removed the is trigger param.

Elastic search looks like it may be more tricky

Anyway if you have ideas to simplify we can discuss. Probably ping you on slack next week

Reply:
Another reason I went with separate tabs is, the handlers, in their reader capacity, support streaming, reading from offset. It gets a little weird to think about what that means and how that should work when you have two different streams/files. And then add in that those files could be local, served, or remote. 
After the process is done it's streigforward enough.  Another part is timestamps... Since we don't necessarily have control over user log format, we might parse incorrectly. 
Neither is necessarily a deal breaker but caused me to explore separate tabs and I was happy with the result so stuck with it.

Reply:
Agree it's very cloce. I think the **only** real difference is the dynamic replacement of the registered handlers and finding the typo of them.

 I **believe** we do not have to do it -  if we make the separate process/thread that  actually directly use already registered handlers - without having to add a wrapper. 

Simply turn the current "wrapper" into 'forwarder" - i.e. the "handler" would only push it to memory queue, and the queue itself would write to regular "airflow.task" without any modifications. 

I have the impression, looking at the code, that the "supports_triggerer" flag is not needed in the handler and if we implement "forwarder" properly, all and every current (streaming) handler of Airflow will work as expected (and the description suggeste it's not the case and more work is needed). 

But maybe I am wrong. and  I do not realize some hidden coupling (i't often when you start actually implementing stuff you find out some hidden couplings implementation details that make simple thing not that simple any more.

Yah. I will discuss on slack .

Reply:
If you have only one handler instance for all triggers how do you ensure that each message goes to its own file?

Reply:
I eat the messages for each trigger each go to their own files?

Reply:
You'll have to remember that The trigger files must be named differently from the task files because when they're uploaded you don't want to overwrite. And we have to remember that the same trigger could run multiple times on different triggerer instances and so we have to name those separately too so that they're not overwriting each other

Reply:
So the handler has to know that this is a trigger writing not a task. That's what is_trigger is doing right now. But I have simplified that and will push that later.

Reply:
> Another reason I went with separate tabs is, the handlers, in their reader capacity, support streaming, reading from offset. It gets a little weird to think about what that means and how that should work when you have two different streams/files. And then add in that those files could be local, served, or remote.

Yeah. Very good point. I have not though about the "offset" - and I guess it it pretty much "logging-system-dependent" and might get wrong when you have two different processes (or even machines) wring in paralell - which indeed has not been the case so far.  So yeah. Agree separate tabs is likely better in this case.

Though I also think (and that's more to @bbovenzi and @pierrejeambrun  - it would be great if we can have the main and triggerer log tab also side-by-side and ideally some way to link scrolling (maybe in the future ) that they could be scrolled through "together". This might get of course extremely tricky and maybe the scrolling part is not needed, but being able to see the task log and triggerer log at least side-by-side with timestamps so that they can be somehow - even manually-corelllated is I think a must. I think having to switch tabs back-forth to see where you deferred and what happened after is pretty terrible UX.

Also I think still the "find and replace handler" if we can avoid it is way better (I will discuss on slack about it - maybe that is not possible/easy to avoid it but it looks pretty "fishy" to me).

Reply:
The "find and replace" is similar to what we do for tasks already. We find the task handler and move it to root. See task command.py. only difference is here I wrap it so we can instantiate one per trigger so each can go to own file. See also my other pr re k8s exec pod stdout. 

I use the supports flag in part because we don't want to take just any handler and wrap. User may have other custom handlers there for other reasons and we leave that alone for their benefit. And we can't just take any fth cus not all have been updated. Also, I wanted a way that we could conditionally disable the pane and provide good signalling to user when their handler does not support individual triggerer logging, and having such a flag makes that straightforward.

Reply:
Just posting here the **summary** of our (long) discussion with @dstandish  that we had on slack - if only for the records but also to track and be able to come back to the discussion in case we decide to improve the solution in the future:

* the current approach is a relatively quick way to solve the painful problem of missing trigger logs so it's ok to go for it now and see if more scalable solution is needed

* it has some limitations that we should be aware of (and make our users aware) - the most important is that it creates multiple (1000s or even 10000s) handlers in a single process - each handler having opened file or socket in parallel might lead to exhaution of sockets (especially in environments like docker and k8s where sockets are actually shared resource). Also each Handlers takes some memory (and for example streaming handlers might take more than "small" amount). Also some streaming handlers might not be capable of runnning multiple clients in the same process.

* this might - in effect - limit Triggerer capabilities in terms of number of paralell tasks handled - and those limits will be system/kernel/deployment dependent. Or it might lead to an excessive memory usage by running Triggerer

* we need to extensively test the resources limits/memory consumption and capabilities of streaming handlers to run multiple clients in the same process before releasing this one to see if these anticipated problems are "real". 

* It's possible that even if we find the limits are there, we decide to relase it and state the limits - warn our users and provide them escape hatches (disabling triggerer logging, proper diagnosis and helpful advices when we hit the resource limits, advising the users to run multiple triggerers etc.)

Also being constructive - I have a possible improvement propoasl  we can implement later if we decide to merge this one that I think is doable and takes into account the way how triggers work. I will describe it in a separate comment/start a separate thread for that.